nat/traversal/pinger.go

```diff
@@ -161,58 +161,33 @@ func (p *Pinger) PingProvider(ip string, localPorts, remotePorts []int, proxyPor
 	return localPort, remotePort, err
 }
 
-func (p *Pinger) ping(conn *net.UDPConn, ttl int, stop <-chan struct{}) error {
-	// Windows detects that 1 TTL is too low and throws an exception during send
-	i := 0
-
+func (p *Pinger) ping(conn *net.UDPConn, remoteAddr *net.UDPAddr, ttl int, stop <-chan struct{}) error {
 	err := ipv4.NewConn(conn).SetTTL(ttl)
 	if err != nil {
 		return errors.Wrap(err, "pinger setting ttl failed")
 	}
 
-	for {
+	for i := 1; time.Duration(i)*p.pingConfig.Interval < p.pingConfig.Timeout; i++ {
 		select {
 		case <-stop:
 			return nil
 		case <-p.stop:
 			return nil
 
 		case <-time.After(p.pingConfig.Interval):
-			log.Debug().Msgf("Pinging %s from %s...", conn.RemoteAddr().String(), conn.LocalAddr().String())
+			log.Debug().Msgf("Pinging %s from %s...", remoteAddr.String(), conn.LocalAddr().String())
 
-			_, err := conn.Write([]byte("continuously pinging to " + conn.RemoteAddr().String()))
+			_, err := conn.WriteToUDP([]byte("continuously pinging to "+remoteAddr.String()), remoteAddr)
 			if err != nil {
 				p.eventPublisher.Publish(event.AppTopicTraversal, event.BuildFailureEvent(StageName, err))
 				return errors.Wrap(err, "pinging request failed")
 			}
-
-			i++
-
-			if time.Duration(i)*p.pingConfig.Interval > p.pingConfig.Timeout {
-				err := errors.New("timeout while waiting for ping ack, trying to continue")
-				p.eventPublisher.Publish(event.AppTopicTraversal, event.BuildFailureEvent(StageName, err))
-				return err
-			}
 		}
 	}
-}
-
-func (p *Pinger) getConnection(ip string, remotePort int, localPort int) (*net.UDPConn, error) {
-	udpAddr, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("%s:%d", ip, remotePort))
-	if err != nil {
-		return nil, err
-	}
-
-	log.Info().Msg("Remote socket: " + udpAddr.String())
-
-	conn, err := net.DialUDP("udp", &net.UDPAddr{Port: localPort}, udpAddr)
-	if err != nil {
-		return nil, err
-	}
 
-	log.Info().Msg("Local socket: " + conn.LocalAddr().String())
-
-	return conn, nil
+	err = errors.New("timeout while waiting for ping ack, trying to continue")
+	p.eventPublisher.Publish(event.AppTopicTraversal, event.BuildFailureEvent(StageName, err))
```

Comment:
It' should be one single result event (success/failure), currently it spams with each parallel ping result

Reply:
Moved it to the top level only.

Reply:
Are you sure, events for single ping case are triggered?

Reply:
There are no single pings any more. Entrypoint is `multiPing` even if there is only single port in the slice.